[FIX] web: stop showing visitors error dialogs for third-party scripts

sdegueldre opened the pull request:
Previously, if an error occurred in a third-party script loaded from a different origin (eg: google analytics or scripts injected into the in-app browsers of Meta apps such as Facebook or Instagram), we would show an error dialog mentioning a "CORS" error with no details.

There are three problems with this: first, this is not actually a CORS error. It is normal error, but in Chrome and Safari, the details of the error are unavailable inside of code from scripts hosted on a different origin, this unavailability is caused by CORS but the error is *not* a CORS error.

The second problem is that we're showing these error dialogs to website visitors, who are both likely to be confused or frustrated by the error and very unlikely to report it.

Lastly, because the error occurred in a third party script, it's likely code that is not even controlled by the Odoo instance in any way, which means that it oftentimes cannot be fixed in Odoo.

In Firefox, the first point doesn't hold true, as Firefox does not hide the error details, but we still do not want to display an error dialog because of the two other points.

This commit fixes this by simply not displaying these errors to visitors outside of debug mode. Actual visitors will not be in debug mode, but it could be useful to still show the error for people developing integrations with third-party libraries and scripts. The error dialog will still not contain much information in non-Firefox browsers, but we encourage the developer to look at the browser console for extra details. We still display these errors for logged-in users in all cases as users are much more likely to report such issues. Ideally, we may not want to show these errors to portal users either, but this information is not currently available in JS.

opw-3857323

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/163710).

rdeodoo commented on addons/web/static/src/core/browser/feature_detection.js:
```diff
@@ -15,6 +15,15 @@ export function isBrowserChrome() {
 }
 
+/**
+ * true if the browser Firefox
```
```suggestion
 * true if the browser is Firefox
```

rdeodoo commented on addons/web/static/src/core/errors/error_service.js:
```diff
@@ -106,14 +109,27 @@ export const errorService = {
                 uncaughtError.traceback = env._t(
-                    `Unknown CORS error\n\n` +
-                        `An unknown CORS error occured.\n` +
-                        `The error probably originates from a JavaScript file served from a different origin.\n` +
-                        `(Opening your browser console might give you a hint on the error.)`
+                    `Third-party script error\n\n` +
+                        `An error whose details are inaccessible to the Odoo framework occurred.\n` +
+                        `This error likely comes from a script served from a different origin.\n` +
+                        `Further information may be available in the browser console.`
```
I wouldn't have changed the wording "just" to make it more "exact" as it will break translations, and I always prefer to keep a "not perfect" but translated wording than fixing it in stable.
Indeed, people might suddenly now see English messages while they don't speak English.

Eg
![image](https://github.com/odoo/odoo/assets/30048408/cbe4fabc-7905-4eae-840f-544645115f2b)

In some main languages it might be re-translated within a few weeks, even in stable, but for some it will never be (the older the version the truer it gets)

rdeodoo replied:
Also, it need to be changed in POT if you still want to change that

rdeodoo reviewed (changes requested):
Nice, thank you :+1: 
Working as expected.

Small typo and I would not change the wording in stable but let me know if you still want to proceed as is for the translated terms and I'll r+ (I would really not change it but.. as you prefer)
And if you keep the wording change, you need to change the .pot

About this comment in the code
`// Firefox doesn't hide details of errors occuring in third-party scripts, check origin explicitly` and this one in the commit message ` but in Chrome and Safari, the details of the error are unavailable inside of code from scripts hosted on a different origin, this unavailability is caused by CORS but the error is not a CORS error.` / `In Firefox, the first point doesn't hold true, as Firefox does not hide the error details,`
Are you sure? Wasn't it supposed to be fixed (as it's a security issue) a loooooooooong time ago? See https://bugzilla.mozilla.org/show_bug.cgi?id=363897

-----------------

For tracking purpose:
The "CORS" error catching was introduced with this [commit](https://github.com/odoo/odoo/commit/4bd8e3e2ed48eb05b9e155826fd47753a85b5731), which came with some useful reading about CORS (third party actually) errors (one is leading to the bugzilla mentioned above). (and it also came with some comment next to those very weird lines which you need explanation to figure (an error without line/col/file looks very strange), but not comment left in code now..)

sdegueldre commented:
> About this comment in the code `// Firefox doesn't hide details of errors occuring in third-party scripts, check origin explicitly` and this one in the commit message ` but in Chrome and Safari, the details of the error are unavailable inside of code from scripts hosted on a different origin, this unavailability is caused by CORS but the error is not a CORS error.` / `In Firefox, the first point doesn't hold true, as Firefox does not hide the error details,` Are you sure? Wasn't it supposed to be fixed (as it's a security issue) a loooooooooong time ago? See https://bugzilla.mozilla.org/show_bug.cgi?id=363897
> 
> For tracking purpose: The "CORS" error catching was introduced with this [commit](https://github.com/odoo/odoo/commit/4bd8e3e2ed48eb05b9e155826fd47753a85b5731), which came with some useful reading about CORS (third party actually) errors (one is leading to the bugzilla mentioned above). (and it also came with some comment next to those very weird lines which you need explanation to figure (an error without line/col/file looks very strange), but not comment left in code now..)

Yes, the reason I added this check is because I tested it on Firefox and you get full details on the error. I did read that issue on bugzilla while investigating, it seems that originally the issue was that you could tell if the user was logged in onto another site by attempting to load that site as a script, as the site would return different content for logged in user resulting in a different parsing error. I think this vulnerability is solved by CORS, as now credentials are no longer included by default. When the attribute `crossorigin="use-credentials"` is used, the server must reply with the appropriate headers (Access-control-allow-origin, which *cannot* be a wild-card asterisk for this, and Access-control-allow-credentials: true), so the third-party origin must opt into this third-party domain being able to read this information very explicitly. I guess that's why Firefox has decided that this error redaction behavior is no longer fully necessary? Either that or it's a bug in Firefox.

I have removed the diff on the error message, I'll add that back in master and probably rename the UncaughtCorsError class as well.

rdeodoo reviewed (approved):
@robodoo r+

Perfect then :+1:

sdegueldre commented:
@rdeodoo I had forgotten to adapt the tests I'm adding after reverting the phrasing changes, should be good now.

rdeodoo commented:
@robodoo r+

rdeodoo commented:
@robodoo retry
Doesn't seem related + branch build green + not much queue + end of the day
